include/can/core/message_handlers/system.hpp

```diff
@@ -56,24 +56,25 @@ class SystemMessageHandler {
   private:
     void visit(std::monostate &) {}
 
-    void visit(DeviceInfoRequest &) {
+    void visit(DeviceInfoRequest &m) {
+        response.message_index = m.message_index;
         writer.send_can_message(can::ids::NodeId::host, response);
     }
 
     void visit(InitiateFirmwareUpdate &) { app_update_start(); }
 
-    void visit(FirmwareUpdateStatusRequest &) {
-        auto status_response =
-            FirmwareUpdateStatusResponse{.flags = app_update_flags()};
+    void visit(FirmwareUpdateStatusRequest &m) {
+        auto status_response = FirmwareUpdateStatusResponse{
```

Comment:
a static member function that builds a response from a request could be helpful here, possibly even one that takes a mostly-premade response so you could use tagged-initialization still. Something like this:

```
struct FirmwareUpdateStatusResponse {
    static auto from_request(FirmwareUpdateStatusResponse premade. FirmwareUpdateStatusRequest request) -> FirmwareUpdateStatusResponse {
        premade.message_id = request.message_id;
        return premade; 
    }
};
```
which could then be called like
```
auto status_response = FirmwareUpdateStatusResponse::from_request({.flags = app_update_flags()}, m);
```
as long as there was a default for the message ids. That way we could establish a pattern of "please build this through the static method" and not have to remember to set the message id.